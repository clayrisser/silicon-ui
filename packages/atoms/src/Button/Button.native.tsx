import React, { FC } from 'react';
import styled, { StyledComponent } from '@emotion/primitives';
import { Button as NativeBaseButton } from 'native-base';
import { TextProps } from 'react-native';
import { useTheme } from 'emotion-theming';
import {
  LayoutProps,
  background,
  border,
  color,
  compose,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import useColor from '../hooks/useColor';
import { Theme } from '../themes';
import { createStyled } from '../styled';
import {
  ButtonProps,
  StyledTextProps,
  antiForwardButtonPropsKeys,
  splitProps
} from './buttonProps';

const StyledText: StyledComponent<
  StyledTextProps,
  TextProps & LayoutProps,
  any
> = styled.Text(compose(color, typography, layout));

const StyledNativeBaseButton = createStyled<ButtonProps>(
  NativeBaseButton,
  [background, border, layout, position, shadow, space],
  antiForwardButtonPropsKeys
);

const Button: FC<ButtonProps> = (props: ButtonProps) => {
  const color = useColor(props);
  const theme: Theme = useTheme();
  const {
    styledButtonProps,
    customButtonProps,
    styledTextProps,
    nativeButtonProps
  } = splitProps({
    ...props,
    color,
    ...(props.backgroundColor !== 'undefined'
      ? {
          backgroundColor:
            theme.colors[props.backgroundColor as string] ||
            props.backgroundColor
        }
      : {})
  });
  const children =
    typeof customButtonProps.children === 'string' ? (
      <StyledText {...styledTextProps} width="100%">
        {customButtonProps.children}
      </StyledText>
    ) : (
      customButtonProps.children
    );
  return (
    <StyledNativeBaseButton {...styledButtonProps} {...nativeButtonProps}>
      {children}
    </StyledNativeBaseButton>
  );
};

Button.defaultProps = {
  // fontFamily: 'body',
  // fontWeight: 'body',
  // lineHeight: 'body',
  autoContrast: false,
  backgroundColor: 'primary',
  children: '',
  fontSize: 2,
  paddingBottom: 2,
  paddingLeft: 2,
  paddingRight: 2,
  paddingTop: 2,
  textAlign: 'center',
  width: '100%'
};

export default Button;
