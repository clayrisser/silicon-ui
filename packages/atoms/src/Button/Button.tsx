import React, { FC } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import {
  background,
  border,
  color,
  compose,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import useColor from '../hooks/useColor';
import {
  ButtonProps,
  DetailedHTMLButtonProps,
  splitProps
} from './buttonProps';

const HTMLButton: StyledComponent<
  DetailedHTMLButtonProps,
  ButtonProps,
  object
> = styled.button(
  compose(
    background,
    border,
    color,
    layout,
    position,
    shadow,
    space,
    typography
  )
);

const Button: FC<ButtonProps> = (props: ButtonProps) => {
  const color = useColor(props);
  const { customButtonProps, styledButtonProps, styledTextProps } = splitProps({
    ...props,
    color
  });

  function handleClick(_e: any) {
    if (props.onPress) props.onPress();
  }

  function handleMouseDown(e: any) {
    if (props.onPressIn) props.onPressIn(e);
  }

  function handleMouseUp(e: any) {
    if (props.onPressOut) props.onPressOut(e);
  }

  return (
    <HTMLButton
      {...styledButtonProps}
      {...styledTextProps}
      {...(customButtonProps as any)}
      onClick={handleClick}
      onMouseDown={handleMouseDown}
      onMouseUp={handleMouseUp}
    />
  );
};

Button.defaultProps = {
  activeOpacity: 0.8,
  autoContrast: false,
  backgroundColor: 'primary',
  borderRadius: 2,
  borderWidth: 0,
  children: '',
  fontFamily: 'body',
  fontSize: 2,
  fontWeight: 'body',
  lineHeight: 'body',
  paddingBottom: 2,
  paddingLeft: 2,
  paddingRight: 2,
  paddingTop: 2,
  width: '100%',
  name: 'first',
  type: 'next',
  className: 'className'
};

export default styled(Button)`
  cursor: pointer;
  font-weight: 500;
  transition-duration: 0.25s;
  transition-property: opacity;
  :active {
    opacity: ${({ activeOpacity }: ButtonProps) => activeOpacity};
  }
`;
