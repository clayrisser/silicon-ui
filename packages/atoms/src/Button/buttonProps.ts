import { DetailedHTMLProps, ButtonHTMLAttributes, ReactNode } from 'react';
import {
  BackgroundColorProps,
  BackgroundProps,
  BorderProps,
  LayoutProps,
  OpacityProps,
  PositionProps,
  ShadowProps,
  SpaceProps,
  TextColorProps,
  TypographyProps
} from 'styled-system';
import { NativeBase } from 'native-base';
import { Theme } from '../themes';

export type DetailedHTMLButtonProps = DetailedHTMLProps<
  ButtonHTMLAttributes<HTMLButtonElement>,
  HTMLButtonElement
>;

export interface StyledButtonProps
  extends BackgroundProps,
    BackgroundColorProps,
    BorderProps,
    LayoutProps,
    OpacityProps,
    PositionProps,
    ShadowProps,
    SpaceProps {}

export interface NativeButtonProps extends Omit<NativeBase.Button, 'color'> {}

export interface StyledTextProps extends TypographyProps, TextColorProps {}

export interface CustomButtonProps {
  autoContrast?: boolean | 'A' | 'AA' | 'AAA';
  children?: ReactNode;
  theme?: Theme;
}

export interface ButtonProps
  extends StyledButtonProps,
    CustomButtonProps,
    StyledTextProps,
    NativeButtonProps {}

export const antiForwardButtonPropsKeys = new Set(['borderColor']);

export const customButtonPropsKeys = new Set(['autoContrast', 'children']);

export const styledTextPropsKeys = new Set([
  'color',
  'fontFamily',
  'fontSize',
  'fontStyle',
  'fontWeight',
  'letterSpacing',
  'lineHeight',
  'textAlign'
]);

export const nativeButtonPropsKeys = new Set([
  'androidRippleColor',
  'badge',
  'dark',
  'disabled',
  'first',
  'full',
  'hasText',
  'icon',
  'iconLeft',
  'iconRight',
  'inputButton',
  'large',
  'last',
  'light',
  'onPress',
  'onPressIn',
  'onPressOut',
  'rounded',
  'small',
  'textStyle',
  'transparent',
  'vertical',
  'block',
  'bordered'
]);

interface SplitProps {
  customButtonProps: CustomButtonProps;
  nativeButtonProps: NativeButtonProps;
  styledButtonProps: StyledButtonProps;
  styledTextProps: StyledTextProps;
}

export function splitProps(props: ButtonProps): SplitProps {
  const customButtonProps: { [key: string]: any } = {};
  const nativeButtonProps: { [key: string]: any } = {};
  const styledButtonProps: { [key: string]: any } = {};
  const styledTextProps: { [key: string]: any } = {};
  Object.entries({ ...props }).forEach(([key, prop]: [string, any]) => {
    if (customButtonPropsKeys.has(key)) {
      customButtonProps[key] = prop;
    } else if (styledTextPropsKeys.has(key)) {
      styledTextProps[key] = prop;
    } else if (nativeButtonPropsKeys.has(key)) {
      nativeButtonProps[key] = prop;
    } else {
      styledButtonProps[key] = prop;
    }
  });
  return {
    customButtonProps,
    nativeButtonProps,
    styledButtonProps,
    styledTextProps
  };
}
