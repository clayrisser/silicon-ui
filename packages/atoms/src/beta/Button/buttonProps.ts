import { NativeBase } from 'native-base';
import { ReactNode } from 'react';
import { SxStyleProp } from 'theme-ui';
import { ThemedStyledProps } from 'native-theme-ui';
import { createSplitProps } from '../../util';

export type ButtonProps = CustomButtonProps &
  StyledNativeBaseButtonProps &
  NativeTextProps;

export interface CustomButtonProps {
  autoContrast?: boolean | 'A' | 'AA' | 'AAA';
  children?: ReactNode;
}

export interface StyledNativeBaseButtonProps
  extends Omit<ThemedStyledProps<NativeBase.Button>, 'ref'> {}
export type StyledNativeBaseButtonSx = Partial<SxStyleProp>;

export interface NativeTextProps {}
export type NativeTextSx = Partial<SxStyleProp>;

interface SplitProps {
  customButtonProps: CustomButtonProps;
  nativeTextProps: NativeTextProps;
  styledNativeBaseButtonProps: StyledNativeBaseButtonProps;
}

export interface SplitSx {
  nativeTextSx: NativeTextSx;
  styledNativeBaseButtonSx: StyledNativeBaseButtonProps;
}

export const splitProps = createSplitProps<ButtonProps, SplitProps, SplitSx>(
  {
    customButtonProps: ['autoContrast', 'children'],
    nativeTextProps: []
  },
  'styledNativeBaseButtonProps',
  { nativeTextSx: /^((color)|(text.+)|(font.+))$/ },
  'styledNativeBaseButtonSx'
);
