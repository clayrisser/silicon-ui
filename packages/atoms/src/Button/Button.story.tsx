import React from 'react';
import { action } from '@storybook/addon-actions';
import { withKnobs, text, number, select } from '@storybook/addon-knobs';
import Button from './Button';
import Wrapper from '../../storybook/Wrapper';
import storiesOf from '../../storybook/storiesOf';
import themes from '../themes';
import withThemesProvider from '../../storybook/withThemesProvider';
// import docs from './Button.docs.mdx';

storiesOf('Button', module)
  .addDecorator(withKnobs)
  .addDecorator(withThemesProvider(themes))
  .addParameters({
    // docs: { page: docs },
    // jest: ['Button.test.tsx']
  })
  .add('with knobs', () => (
    <Wrapper>
      <Button
        backgroundColor={text('backgroundColor', 'primary')}
        borderRadius={number('borderRadius', 2)}
        onFocus={action('onFocus')}
        onPress={action('onPress')}
        onPressIn={action('onPressIn')}
        onPressOut={action('onPressOut')}
        autoContrast={select<'A' | 'AA' | 'AAA'>(
          'autoContrast',
          {
            false: '' as 'A',
            A: 'A',
            AA: 'AA',
            AAA: 'AAA'
          },
          'AA'
        )}
      >
        {text('children', 'click me')}
      </Button>
    </Wrapper>
  ));
