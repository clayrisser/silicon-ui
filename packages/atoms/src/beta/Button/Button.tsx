import React, { FC } from 'react';
import { NativeBase, Button as NativeBaseButton } from 'native-base';
import { styled, Text as NativeText } from 'native-theme-ui';
import { SxStyleProp } from 'theme-ui';
import { ButtonProps, splitProps } from './buttonProps';

const StyledNativeBaseButton = styled<NativeBase.Button>(NativeBaseButton)({});

const defaultSx: SxStyleProp = {
  backgroundColor: 'primary',
  color: 'text',
  position: 'inherit',
  textAlign: 'center'
};

const Button: FC<ButtonProps> = (props: ButtonProps) => {
  const {
    customButtonProps,
    nativeTextSx,
    sx,
    themedNativeBaseButtonProps
  } = splitProps(props, defaultSx);

  function renderChildren() {
    const { children } = customButtonProps;
    if (!children) return null;
    if (typeof children === 'string') {
      return (
        <NativeText sx={{ ...nativeTextSx, width: '100%' }}>
          {children}
        </NativeText>
      );
    }
    return children;
  }

  return (
    <StyledNativeBaseButton {...themedNativeBaseButtonProps} sx={sx}>
      {renderChildren()}
    </StyledNativeBaseButton>
  );
};

Button.defaultProps = {};

export default Button;
