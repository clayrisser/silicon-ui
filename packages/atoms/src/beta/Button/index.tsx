import Button from './Button';
import { ButtonProps } from './buttonProps';

export { ButtonProps };
export default Button;
