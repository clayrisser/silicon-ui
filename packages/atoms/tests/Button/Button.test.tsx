import React from 'react';
import renderer, { act, create } from 'react-test-renderer';

import { render, fireEvent } from '@testing-library/react-native';
// import { Button, Text, TextInput, View } from 'react-native';
// import { render, fireEvent } from '@testing-library/react';

import Button from '../../src/Button/Button';

describe('<Button />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Button />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('<Button withProps /> ', () => {
  it('renders with disabled prop', () => {
    const borderRadius = 2;
    const borderWidth = 2;
    const backgroundColor = 'free';
    const fontSize = '';
    const fontWeight = '';
    const fontFamily = '';
    const width = 3;

    const tree = renderer
      .create(
        <Button
          borderRadius={borderRadius}
          backgroundColor={backgroundColor}
          borderWidth={borderWidth}
          fontSize={fontSize}
          fontWeight={fontWeight}
          fontFamily={fontFamily}
          width={width}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('<Button />', () => {
  it('handles onPress', () => {
    const onPress = jest.fn();
    const onMouseDown = jest.fn();
    const onMouseUp = jest.fn();
    // const { container } = render(<Button onPress={onPress} />);
    // fireEvent.press(container.querySelector('HTMLButton')!);
    // expect(onPress).toHaveBeenCalledTimes(1);
    // const handleClick = jest.fn();
    // const { getByText } = render(<Button onPress={onPress} />);
    // const button = getByText('click me');
    // fireEvent.press(button);
    // expect(onPress).toBe(undefined);
    let component: any;
    act(() => {
      component = create(
        <Button
          onPress={onPress}
          onPressIn={onMouseDown}
          onPressOut={onMouseUp}
        />
      );
    });
    const rootInstance: any = component.root;
    const button = rootInstance.findAllByType(Button);
    expect(button.length).toBe(1);
    //  act(() => button[0].props.onPress());
    expect(button[0].props.onPress()).toBe(undefined);
    expect(button[0].props.onPressIn()).toBe(undefined);
    expect(button[0].props.onPressOut()).toBe(undefined);
  });

  // it('handles onPress', () => {
  //   const handleClick = jest.fn();
  //   const { container } = render(<Button onPress={handleClick} />);
  //   fireEvent.click(container.querySelector('HTMLButton')!);
  //   expect(handleClick).toHaveBeenCalledTimes(1);
  // });
});
