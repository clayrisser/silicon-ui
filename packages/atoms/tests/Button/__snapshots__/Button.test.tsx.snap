// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders correctly 1`] = `
<button
  className="css-1k7x4jn"
  fontFamily="body"
  fontSize={0}
  fontWeight="body"
  onClick={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  width="100%"
>
  
</button>
`;

exports[`<Button withProps />  renders with disabled prop 1`] = `
<button
  className="css-1nh6o6b"
  fontFamily=""
  fontSize=""
  fontWeight=""
  onClick={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  width={3}
>
  
</button>
`;
